text/0062-consistent-behavior-for-video-sources-that-do-not-have-video.md

```diff
@@ -0,0 +1,95 @@
+```c
+#define OBS_SOURCE_CAP_MIGHT_NOT_HAVE_VIDEO (1 << 17)
+```
+Setting this flag requires that the source implements and respects the other API's shown below.
+Any source that might not have video should set this.
+The idea of a new flag is that not all sources (e.g., text sources or color sources) require such a setting, so they should not get a UI for this or need to implement new APIs.
+
+
+We create an enum to for the available states, as well as methods to get this from and set it on the source.
+```c
+enum obs_source_no_video_behavior {
+    OBS_SOURCE_NO_VIDEO_BEHAVIOR_SHOW_BLACK_FRAME,
+    OBS_SOURCE_NO_VIDEO_BEHAVIOR_SHOW_NOTHING,
+    OBS_SOURCE_NO_VIDEO_BEHAVIOR_SHOW_LAST_FRAME,
+}
+enum obs_source_no_video_behavior obs_source_get_no_video_behavior(obs_source_t *source);
+void obs_source_set_no_video_behavior(obs_source_t *source, enum obs_source_no_video_behavior behavior);
+```
+
+`obs_source_info` gets extended with a `has_video` callback (see "Synchronous sources" for context):
+```c
+struct obs_source_info {
+/* ... */
+    bool (*has_video)(void *data);
+};
+```
+
+### Source Behavior
+#### Async sources
+If an async source at some point no longer has new video to output (e.g., because a Video Capture Device gets disconnected), it should call `obs_source_output_video(source, NULL)`.
+Note that passing `NULL` as the parameter no longer means that no video will be shown - libobs will handle the actual video output.
+
+Asynchronous sources should not call `obs_source_output_video(source, NULL)` if they do not set the `OBS_SOURCE_CAP_MIGHT_NOT_HAVE_VIDEO` flag.
+Doing so results in undefined behavior.
+
+#### Synchronous sources
+Synchronous sources that have the `OBS_SOURCE_CAP_MIGHT_NOT_HAVE_VIDEO` flag set must implement the `has_video` callback.
+This will get called before rendering the source.
+If the source does not have video, it should return `false` in `has_video`.
+In this case, `video_render` will not get called; instead libobs will handle the video that gets shown.
+
+### UI
+A final UI for this should be decided by dedicated UI/UX discussions, but here are suggestions:
+
+Ideally, this is part of a tabbed properties system ([example](https://github.com/obsproject/obs-studio/pull/8155)).
+If a source has set the `OBS_SOURCE_CAP_MIGHT_NOT_HAVE_VIDEO` flag, it would show one combobox or radio button with the three options:
+```
+When no video is shown:
+- Show a black image
+- Show nothing
+- Show the last frame
+```
+
+If tabbed properties are not an option, the options could be radio button added at the bottom of the source's properties.
+However, it would not technically be a property and not set any "normal" source settings (though to the user it would look like a normal property).
+
+Showing the black image is the default behavior, and is also what gets returned by `obs_source_get_no_video_behavior` by default.
```

Comment:
What resolution would this source get? I also imagine this being an unfavorable scenario, where you might be streaming and want to add a source (for example, an image source). You probably wouldn't want your stream to immediately go black until you configure some kind of setting that makes actual content appear.

Reply:
> I also imagine this being an unfavorable scenario, where you might be streaming and want to add a source (for example, an image source). You probably wouldn't want your stream to immediately go black until you configure some kind of setting that makes actual content appear.

*Assuming* it is agreed that unconfigured sources (e.g. Image Sources with an image path explicitly cleared, see https://github.com/obsproject/rfcs/pull/62/files#r1729999264) do not go into the "no video" state, I don't think that should pose an issue. A source that's newly added would have a cleared file path or similar, meaning that it wouldn't trigger this.

> What resolution would this source get?

I don't think there is a correct answer for this since (since there can be multiple canvases, etc), I'd go for some arbitrary value (e.g 800x800). I can't of the top of my head think of a source that would go into "no video" on initial creation (see above), and as soon as it has video at some point that resolution will get remembered.